File: sdk/validation.py
"""Validation helpers for PFT Task Node SDK."""

from __future__ import annotations

import re


CID_PREFIXES = ("bafk", "bafy")
MAX_INT32 = 2**31 - 1


class ValidationError(ValueError):
    """Raised when inputs fail validation."""


def validate_nonempty_content(content: bytes) -> None:
    if not isinstance(content, (bytes, bytearray)) or len(content) == 0:
        raise ValidationError("content must be non-empty bytes")


def validate_cid(cid: str) -> None:
    if not isinstance(cid, str) or not cid:
        raise ValidationError("cid must be a non-empty string")
    if not cid.startswith(CID_PREFIXES) or len(cid) < 20:
        raise ValidationError(f"cid has unexpected format: {cid}")


def validate_xrp_address(address: str) -> None:
    if not isinstance(address, str) or not address:
        raise ValidationError("address must be a non-empty string")
    if not re.fullmatch(r"r[1-9A-HJ-NP-Za-km-z]{24,34}", address):
        raise ValidationError(f"invalid XRP address format: {address}")


def validate_non_negative_int(value: int, name: str) -> None:
    if not isinstance(value, int) or value < 0:
        raise ValidationError(f"{name} must be a non-negative integer")
    if value > MAX_INT32:
        raise ValidationError(f"{name} must be <= {MAX_INT32}")


def validate_url(url: str) -> None:
    if not isinstance(url, str) or not url:
        raise ValidationError("url must be a non-empty string")
    if not (url.startswith("http://") or url.startswith("https://")):
        raise ValidationError(f"url must start with http:// or https://: {url}")

File: ts/src/validation.ts
const CID_PREFIXES = ["bafk", "bafy"] as const;

export class ValidationError extends Error {}

export function validateNonEmptyBytes(bytes: Uint8Array) {
  if (!(bytes instanceof Uint8Array) || bytes.length === 0) {
    throw new ValidationError("payload must be non-empty bytes");
  }
}

export function validateCid(cid: string) {
  if (!cid || typeof cid !== "string") {
    throw new ValidationError("cid must be a non-empty string");
  }
  if (!CID_PREFIXES.some((p) => cid.startsWith(p)) || cid.length < 20) {
    throw new ValidationError(`cid has unexpected format: ${cid}`);
  }
}

export function validateXrpAddress(address: string) {
  if (!address || typeof address !== "string") {
    throw new ValidationError("address must be a non-empty string");
  }
  if (!/^r[1-9A-HJ-NP-Za-km-z]{24,34}$/.test(address)) {
    throw new ValidationError(`invalid XRP address format: ${address}`);
  }
}

export function validateNonNegativeInt(value: number, name: string) {
  if (!Number.isInteger(value) || value < 0) {
    throw new ValidationError(`${name} must be a non-negative integer`);
  }
  if (value > 2 ** 31 - 1) {
    throw new ValidationError(`${name} must be <= ${2 ** 31 - 1}`);
  }
}
